OSD-16162: Add Backplane Session command

samanthajayasinghe opened the pull request:
### What type of PR is this?

feature

### What this PR does / Why we need it?
#### Create a New Backplane session 
```
ocm backplane session test-session -c <cluster-id>
Setting up environment...
INFO[0002] Target cluster                                ID=<cluster-id>
[test-session (<cluster-info-PS1>)]$
```
you can directly login via cluster id 

```
ocm backplane session <cluster-id>
Setting up environment...
INFO[0002] Target cluster                                ID=<cluster-id>
[test-session (<cluster-info-PS1>)]$
```

#### Delete backplane session 
```
ocm backplane  session --delete test-session 
Cleaning up Backplane session test-session
```

#### Get cluster session history 
```
[ <session-alias-orcluster-id> (<cluster-info-PS1>]$ history 
    1  2023-05-08 15:06:05 oc get nodes
    2  2023-05-08 15:06:13 oc get co
    3  2023-05-08 15:06:40 history 
```

### How to change backplane session save path 
By default all session will save into ~/backplane directory 
You can change the dir path via config 
```
{
   "url": "your-bp-url"
    "proxy-url": "your-proxy-url"
   "session-dir":"your-session-dir"
}
```

### Which Jira/Github issue(s) does this PR fix?

_Resolves #_
https://issues.redhat.com/browse/OSD-16162
### Special notes for your reviewer

### Pre-checks (if applicable)

- [x] Ran unit tests locally
- [x] Validated the changes in a cluster
- [ ] Included documentation changes with PR

samanthajayasinghe commented:
/test all

samanthajayasinghe commented:
/test all

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+	}(scriptfile)
+	_, err = scriptfile.WriteString(content)
+	if err != nil {
+		return fmt.Errorf("error writing to file %s: %v", path, err)
```
Capitalise?

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+	}
+	err = os.Chmod(path, 0700)
+	if err != nil {
+		return fmt.Errorf("can't update permissions on file %s: %v", path, err)
```
Capitalise?

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+	if _, err := os.Stat(filename); errors.Is(err, os.ErrNotExist) {
+		file, err = os.Create(filename)
+		if err != nil {
+			return nil, fmt.Errorf("can't create file %s: %v", filename, err)
```
Capitalise?

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+		// Setting up the flags
+		err := login.LoginCmd.Flags().Set("multi", "true")
+		if err != nil {
+			return fmt.Errorf("error occered when setting multi flag %v", err)
```
Capitalise?

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+		// Setting up the flags
+		err := login.LoginCmd.Flags().Set("multi", "true")
+		if err != nil {
+			return fmt.Errorf("error occered when setting multi flag %v", err)
```
Capitalise?

samanthajayasinghe replied:
[ST1005 ](https://staticcheck.io/docs/checks#ST1005) rules

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+		}
+		err = login.LoginCmd.Flags().Set("kube-path", e.Path)
+		if err != nil {
+			return fmt.Errorf("error occered when kube-path flag %v", err)
```
Capitalise?

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+		// Execute login command
+		err = login.LoginCmd.RunE(cmd, []string{e.Options.ClusterId})
+		if err != nil {
+			return fmt.Errorf("error occered when login to the cluster %v", err)
```
Capitalise?

samanthajayasinghe replied:
[ST1005 ](https://staticcheck.io/docs/checks#ST1005) rules

hectorakemp commented on pkg/cli/session/session_test.go:
```diff
@@ -0,0 +1 @@
+package session
```
Probably need some tests in here @samanthajayasinghe ?

samanthajayasinghe replied:
I'll add a unit test via sub ticket since I'm planing to cover 80% this code

hectorakemp replied:
:+1:

hectorakemp commented on pkg/info/info.go:
```diff
@@ -17,6 +17,9 @@ const (
 	BACKPLANE_CONFIG_DEFAULT_FILE_NAME = "config.json"
 
+	// Session
+	BACKPLANE_SESSION_DEFAULT_FILE_PATH = "backplane"
```
As mentioned before, can we make this configurable in `.config/backplane/config.json`?

samanthajayasinghe replied:
Done

hectorakemp reviewed (changes requested):
Overall, great work and a big backplane milestone that we're getting closer to! I've suggested a few small changes that I think we should address before we merge.
 
I found it difficult to get it working and had some issues with returning to my previous session. I wasn't sure exactly what I was looking for. It might be good to get a detailed test plan of what to look for when trying out this feature.

```
See below logs: 
[17:07:30] hectorkemp:backplane-cli git:(session) $ ./ocm-backplane session test-session -c 23j54t6l69iqi76a6jjcs6st4cd3f6i5
Setting up environment...
INFO[0001] Target cluster                                ID=23j54t6l69iqi76a6jjcs6st4cd3f6i5 Name=hkemp-412-hive
Switching to Backplane session test-session
[17:07:47] hectorkemp:test-session $ oc get po
No resources found in default namespace.

[17:12:19] hectorkemp:backplane-cli git:(session) $ ./ocm-backplane session test-session
Setting up environment...
Switching to Backplane session test-session
[17:12:23] hectorkemp:test-session $ oc get po
W0509 17:12:26.632177  338934 loader.go:221] Config not found: /home/hectorkemp/backplane/test-session/23j54t6l69iqi76a6jjcs6st4cd3f6i5/config
```

samanthajayasinghe commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+	if e.Options.ClusterId == "" && e.Options.Alias == "" {
+		err := cmd.Help()
+		if err != nil {
+			return fmt.Errorf("could not print help")
```
According to ST1005 guidelines, an error string shouldn't capitalize. 
https://staticcheck.io/docs/checks#ST1005

samanthajayasinghe commented on cmd/ocm-backplane/session/session.go:
```diff
@@ -0,0 +1,71 @@
+		"reset",
+		"r",
+		true,
+		"Reset session",
```
Removed `reset` option seems it make no sense

samanthajayasinghe commented:
> Overall, great work and a big backplane milestone that we're getting closer to! I've suggested a few small changes that I think we should address before we merge.
> 
> I found it difficult to get it working and had some issues with returning to my previous session. I wasn't sure exactly what I was looking for. It might be good to get a detailed test plan of what to look for when trying out this feature.
> 
> See below logs:
> 
> ```
> [17:07:30] hectorkemp:backplane-cli git:(session) $ ./ocm-backplane session test-session -c 23j54t6l69iqi76a6jjcs6st4cd3f6i5
> Setting up environment...
> INFO[0001] Target cluster                                ID=23j54t6l69iqi76a6jjcs6st4cd3f6i5 Name=hkemp-412-hive
> Switching to Backplane session test-session
> [17:07:47] hectorkemp:test-session $ oc get po
> No resources found in default namespace.
> 
> [17:12:19] hectorkemp:backplane-cli git:(session) $ ./ocm-backplane session test-session
> Setting up environment...
> Switching to Backplane session test-session
> [17:12:23] hectorkemp:test-session $ oc get po
> W0509 17:12:26.632177  338934 loader.go:221] Config not found: /home/hectorkemp/backplane/test-session/23j54t6l69iqi76a6jjcs6st4cd3f6i5/config
> ```

Added some help text when login to the session

samanthajayasinghe commented on cmd/ocm-backplane/login/login.go:
```diff
@@ -78,6 +79,13 @@ func init() {
+		&args.kubeConfigPath,
+		"kube-path",
+		"",
+		"Save kube configuration in specific path",
```
Added some validation with errors 
https://github.com/openshift/backplane-cli/blob/10f945a80a7c168c17ea285b9898c9c132a6aeb9/cmd/ocm-backplane/login/login.go#L163

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Dir = e.Path
+	_ = cmd.Run() // add error checking
```
@samanthajayasinghe does this need an error check?

hectorakemp commented on pkg/cli/session/session.go:
```diff
@@ -0,0 +1,408 @@
+			if errors.Is(err, os.ErrNotExist) {
+				fmt.Println("Nothing to kill")
+			}
+			//return fmt.Errorf("failed to read file .killpids: %v", err)
```
@samanthajayasinghe do you want to delete this comment?

hectorakemp reviewed (changes requested):
Just the two comments I've bumped, otherwise looks great (given that we'll do tests in a separate PR).  I think once we add a description of how to use the command in the readme (like the other commands) then it's good to go!

supreeth7 commented on cmd/ocm-backplane/session/session.go:
```diff
@@ -0,0 +1,71 @@
+	}
+	sessionCmd := &cobra.Command{
+		Use:               "session [flags] [session-alias]",
+		Short:             "Create an environment to interact with a cluster",
```
+1 for added context, instead of doing it in code @samanthajayasinghe can you update the ReadMe by adding a section the purpose of this command?

samanthajayasinghe commented:
> Just the two comments I've bumped, otherwise looks great (given that we'll do tests in a separate PR). I think once we add a description of how to use the command in the readme (like the other commands) then it's good to go!

Thanks for reviewing @hectorakemp.
Addressed most of the review comments and added a readme section too.

https://github.com/openshift/backplane-cli/blob/2f04f7f6904501214c2de1110c7968210f09b0d3/README.md?plain=1#L186

hectorakemp commented:
/lgtm

supreeth7 reviewed (approved):
/lgtm
/approve
